File: frontend/src/types/invoice.ts
export interface InvoiceItem {
    id: number;
    invoice_id: number;
    stock_item_id: number;
    quantity: number;
    rate: number;
    tax_amount: number;
    discount_amount: number;
    line_total: number;
}

export interface Invoice {
    id: number;
    company_id: number;
    customer_id: number;
    sale_id?: number;
    invoice_number: string;
    invoice_date: string;
    due_date?: string;
    total_amount: number;
    status: 'Paid' | 'Unpaid' | 'Cancelled';
    notes?: string;
    items: InvoiceItem[];
    created_at: string;
    updated_at: string;
}

export interface InvoiceItemCreate {
    stock_item_id: number;
    quantity: number;
    rate: number;
    tax_amount?: number;
    discount_amount?: number;
    line_total: number;
}

export interface InvoiceCreate {
    customer_id: number;
    sale_id?: number;
    invoice_number: string;
    invoice_date: string;
    due_date?: string;
    total_amount: number;
    status?: 'Paid' | 'Unpaid' | 'Cancelled';
    notes?: string;
    items: InvoiceItemCreate[];
}

export interface InvoiceListParams {
    company_id: number;
    limit?: number;
    offset?: number;
}

File: backend/app/schemas/invoice.py
from decimal import Decimal
from datetime import date
from pydantic import BaseModel, ConfigDict
from typing import Optional, List
from app.models.enums import InvoiceStatusEnum

class InvoiceItemBase(BaseModel):
    stock_item_id: int
    quantity: Decimal
    rate: Decimal
    tax_amount: Decimal = Decimal('0.00')
    discount_amount: Decimal = Decimal('0.00')
    line_total: Decimal

class InvoiceItemCreate(InvoiceItemBase):
    pass

class InvoiceItemResponse(InvoiceItemBase):
    id: int
    invoice_id: int
    
    model_config = ConfigDict(from_attributes=True)

class InvoiceBase(BaseModel):
    invoice_number: str
    invoice_date: date
    due_date: Optional[date] = None
    customer_id: int
    sale_id: Optional[int] = None
    total_amount: Decimal
    status: InvoiceStatusEnum = InvoiceStatusEnum.UNPAID
    notes: Optional[str] = None

class InvoiceCreate(InvoiceBase):
    items: List[InvoiceItemCreate]

class InvoiceUpdate(BaseModel):
    invoice_number: Optional[str] = None
    invoice_date: Optional[date] = None
    due_date: Optional[date] = None
    customer_id: Optional[int] = None
    sale_id: Optional[int] = None
    total_amount: Optional[Decimal] = None
    status: Optional[InvoiceStatusEnum] = None
    notes: Optional[str] = None
    items: Optional[List[InvoiceItemCreate]] = None

class InvoiceResponse(InvoiceBase):
    id: int
    company_id: int
    items: List[InvoiceItemResponse]
    
    model_config = ConfigDict(from_attributes=True)
